synapse/cortex.py

```diff
@@ -6179,6 +6199,557 @@ async def iterTagPropRows(self, layriden, tag, prop, form=None, stortype=None, s
+
+            if user.iden != iden and not user.isAdmin():
+                mesg = f'User {user.name} cannot create vaults for user {_user.name}'
+                raise s_exc.AuthDeny(mesg=mesg)
+
+            # The user is the admin, everyone else is denied
+            await self._setEasyPerm(vault, 'users', iden, s_cell.PERM_ADMIN)
+
+        elif scope == 'role':
+            user.confirm(('vaults', 'role', 'add'), default=True) # TODO: Remove default=True when SYN-6097 is resolved.
+
+            role = self.auth.role(iden)
+            if role is None:
+                raise s_exc.NoSuchRole(mesg=f'Role with iden {iden} not found')
+
+            if not user.hasRole(iden) and not user.isAdmin():
+                mesg = f'User {user.name} cannot create vaults for role {role.name}'
+                raise s_exc.AuthDeny(mesg=mesg)
+
+            # The creator is the admin, members of the role can edit, everyone else is denied
+            await self._setEasyPerm(vault, 'users', user.iden, s_cell.PERM_ADMIN)
+            await self._setEasyPerm(vault, 'roles', iden, s_cell.PERM_EDIT)
+
+        else:
+            user.confirm(('vaults', 'add'), default=True) # TODO: Remove default=True when SYN-6097 is resolved.
+
+            await self._setEasyPerm(vault, 'users', user.iden, s_cell.PERM_ADMIN)
+
+        if iden and self._getVaultByTSI(vtype, scope, iden) is not None:
+            raise s_exc.DupName(mesg=f'Vault already exists')
+
+        if self._getVaultByName(name) is not None:
+            raise s_exc.DupName(mesg=f'Vault {name} already exists')
+
+        return await self._push('cortex:vault:set', viden, vault)
+
+    @s_nexus.Pusher.onPushAuto('cortex:vault:set')
+    async def _addVault(self, iden, vault):
+        self._setVault(iden, vault)
+        return iden
+
+    async def setVaultData(self, name, key, valu, user=None):
+        '''
+        Set vault data.
+
+        This function sets the `key`:`valu` into the vault.
+
+        Permissions: `user` must have at least `PERM_EDIT` on the vault.
+
+        Args:
+            name (str): The name or iden of the vault to edit.
+            key (str): Vault data key.
+            valu (str): Vault data value.
+            user (Optional[HiveUser]): User trying to edit the vault data.
+
+        Raises:
+            synapse.exc.NoSuchIden: Vault with `name` (name or iden) does not exist.
+            synapse.exc.AuthDeny: `user` does not have permission to edit vault.
+
+        Returns: (boolean) True if data was set, false otherwise.
+        '''
+        try:
+            s_msgpack.en({key: valu})
+        except s_exc.NotMsgpackSafe as exc:
+            raise s_exc.BadArg(mesg=f'Vault data must be msgpack safe.') from None
+
+        vault = self._getVaultByName(name, throw=True)
+
+        if user is None:
+            user = self.auth.rootuser
+
+        self._reqEasyPerm(vault, user, s_cell.PERM_EDIT)
+
+        if valu is None:
```

Comment:
A user can create a `data` value as a dictionary with null items in it. If we allow that, we should allow setting a null item. 

We can use the `s_common.novalu` singleton to mark this for popping a value instead. From a Storm library perspective, if a user passed in `$lib.undef` as the value, we could then map that singleton to `s_common.novalu` when calling this API?